homeassistant/components/esphomelib/__init__.py

```diff
@@ -0,0 +1,338 @@
+
+    hass.data.setdefault(DOMAIN, {})
+
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
+    password = entry.data[CONF_PASSWORD]
+
+    cli = APIClient(hass.loop, host, port, password)
+    await cli.start()
+
+    async def on_stop(event):
+        """Cleanup the socket client on HA stop."""
+        await _cleanup_instance(hass, entry)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_stop)
+
+    # Store client in per-config-entry hass.data
+    entry_data = get_entry_data(hass, entry)
+    entry_data[DATA_CLIENT] = cli
+    entry_data[DATA_AVAILABLE] = False
+
+    try_connect = await _setup_auto_reconnect_logic(
+        hass, cli, entry.entry_id, host)
+
+    @callback
+    def async_on_state(state):
+        """Send dispatcher updates when a new state is received."""
+        async_dispatcher_send(
+            hass, DISPATCHER_ON_STATE.format(entry.entry_id), state)
+
+    async def on_login():
+        """Subscribe to states and list entities on successful API login."""
+        try:
+            entry_data[DATA_DEVICE_INFO] = await cli.device_info()
+            entry_data[DATA_AVAILABLE] = True
+
+            entity_infos = await cli.list_entities()
+            async_dispatcher_send(
+                hass, DISPATCHER_ON_LIST.format(entry.entry_id),
+                entity_infos)
+            await cli.subscribe_states(async_on_state)
+        except APIConnectionError as err:
+            _LOGGER.warning("Error getting initial data: %s", err)
+            # Re-connection logic will trigger after this
+            await cli.disconnect()
+
+    cli.on_login = on_login
+
+    # This is a bit of a hack: We schedule complete_setup into the
+    # event loop and return immediately (return True)
+    #
+    # Usually, we should avoid that so that HA can track which components
+    # have been started successfully and which failed to be set up.
+    # That doesn't work here for two reasons:
+    #  - We have our own re-connect logic
+    #  - Before we do the first try_connect() call, we need to make sure
+    #    all dispatcher event listeners have been connected, so
+    #    async_forward_entry_setup needs to be awaited. However, if we
+    #    would await async_forward_entry_setup() in async_setup_entry(),
+    #    we would end up with a deadlock.
+    #
+    # Solution is: complete the setup outside of the async_setup_entry()
+    # function. HA will wait until the first connection attempt is made
+    # before starting up (as it should), but if the first connection attempt
+    # fails we will schedule all next re-connect attempts outside of the
+    # tracked tasks (hass.loop.create_task). This way HA won't stall startup
+    # forever until a connection is successful.
+
+    async def complete_setup():
+        """Complete the config entry setup."""
+        for component in HA_COMPONENTS:
+            await hass.config_entries.async_forward_entry_setup(
```

Comment:
Store the tasks in a list and use `asyncio.wait`, since order isn't important here.